File: CrossfitDiary/CrossfitDiary.Web/Scripts/Pages/Models/ViewModels/ToLogWorkoutViewModel.ts
module Models {
  import Serializable = General.Serializable;

  interface IToLogWorkoutViewModel {
    crossfitterWorkoutId: number;
    date: string;
    partialRepsFinished: number;
    roundsFinished: number;
    repsToFinishOnCapTime: number;
    selectedWorkoutId: number;
    timePassed: string;
    isEditMode: boolean;
    canBeRemovedByCurrentUser?: boolean;
    workouterName?: string;
    workouterId?: string;
    displayDate?: string;
    workoutViewModel?: WorkoutViewModel;
  }
  export class ToLogWorkoutViewModel implements Serializable<ToLogWorkoutViewModel>{

    crossfitterWorkoutId: number;
    date: string;
    partialRepsFinished?: number;
    roundsFinished?: number;
    repsToFinishOnCapTime?: number;
    selectedWorkoutId: number;
    timePassed: string;
    isEditMode: boolean;
    canBeRemovedByCurrentUser?: boolean;
    workouterName?: string;
    workouterId?: string;
    displayDate?: string;
    workoutViewModel?: WorkoutViewModel;

    constructor(
      params?: IToLogWorkoutViewModel) {
      if (params == null) {
        return;
      }
      this.crossfitterWorkoutId = params.crossfitterWorkoutId;
      this.date = params.date;
      this.partialRepsFinished = params.partialRepsFinished;
      this.roundsFinished = params.roundsFinished;
      this.selectedWorkoutId = params.selectedWorkoutId;
      this.timePassed = params.timePassed;
      this.isEditMode = params.isEditMode;
      this.repsToFinishOnCapTime = params.repsToFinishOnCapTime;
      this.canBeRemovedByCurrentUser = params.canBeRemovedByCurrentUser;
      this.workouterName = params.workouterName;
      this.workouterId = params.workouterId;
      this.displayDate = params.displayDate;
      this.workoutViewModel = params.workoutViewModel;
    }

    deserialize(jsonInput): ToLogWorkoutViewModel {
      if (jsonInput == null) {
        return null;
      }

      return new ToLogWorkoutViewModel({
        crossfitterWorkoutId: jsonInput.crossfitterWorkoutId,
        date: jsonInput.date,
        isEditMode: jsonInput.isEditMode,
        partialRepsFinished: jsonInput.partialRepsFinished,
        roundsFinished: jsonInput.roundsFinished,
        selectedWorkoutId: jsonInput.selectedWorkoutId,
        timePassed: jsonInput.timePassed,
        repsToFinishOnCapTime: jsonInput.repsToFinishOnCapTime,
        canBeRemovedByCurrentUser: jsonInput.canBeRemovedByCurrentUser,
        workouterName: jsonInput.workouterName,
        workouterId: jsonInput.workouterId,
        displayDate: jsonInput.displayDate,
        workoutViewModel: new WorkoutViewModel().deserialize(jsonInput.workoutViewModel)
      });
    }

  }
}

File: CrossfitDiary/CoreApp/CrossfitDiaryCore.Web/ViewModels/ToLogWorkoutViewModel.cs
using System;
using Newtonsoft.Json;

namespace CrossfitDiaryCore.Web.ViewModels
{
    public class ToLogWorkoutViewModel
    {
        [JsonProperty("id")]
        public int Id { get; set; }

        [JsonProperty("canBeRemovedByCurrentUser")]
        public bool CanBeRemovedByCurrentUser { get; set; }

        [JsonProperty("crossfitterWorkoutId")]
        public int CrossfitterWorkoutId { get; set; }

        [JsonConverter(typeof(DateFormatConverter), "dd.MM.yyyy")]
        public DateTime Date { get; set; }

        [JsonProperty("displayDate")]
        public string DisplayDate { get; set; }

        [JsonProperty("distance")]
        public int? Distance { get; set; }

        [JsonProperty("isEditMode")]
        public bool IsEditMode { get; set; }

        [JsonProperty("isPlanned")]
        public bool IsPlanned { get; set; }

        [JsonProperty("planningWorkoutLevel")]
        public PlanningWorkoutLevel? PlanningWorkoutLevel { get; set; }

        [JsonProperty("isRx")]
        public bool IsRx { get; set; }

        [JsonProperty("partialRepsFinished")]
        public int? PartialRepsFinished { get; set; }

        [JsonProperty("repsToFinishOnCapTime")]
        public int? RepsToFinishOnCapTime { get; set; }

        [JsonProperty("roundsFinished")]
        public int? RoundsFinished { get; set; }

        [JsonProperty("selectedWorkoutId")]
        public int SelectedWorkoutId { get; set; }

        // TODO: Move to separate view model?
        [JsonProperty("selectedWorkoutName")]
        public string SelectedWorkoutName { get; set; }

        [JsonProperty("timePassed")]
        public string TimePassed { get; set; }


        [JsonProperty("wasFinished")]
        public bool WasFinished { get; set; }

        [JsonProperty("workouterId")]
        public string WorkouterId { get; set; }

        [JsonProperty("workouterName")]
        public string WorkouterName { get; set; }

        [JsonProperty("workoutViewModel")]
        public WorkoutViewModel WorkoutViewModel { get; set; }

    }

    public enum PlanningWorkoutLevel
    {
        Scaled = 0,
        Rx = 1,
        RxPlus = 2
    }
}